File: EPI/14.5/swift/14.5.swift
public typealias Event = (start: Int, end: Int)

typealias Boundary = (time: Int, type: BoundaryType)

enum BoundaryType {
    case Start
    case End
}

public func maxConcurrentEvents(_ events: [Event]) -> Int {
    let boundaries = events
        .flatMap { event -> [Boundary] in [
            (time: event.start, type: .Start),
            (time: event.end, type: .End)
        ]}
        .sorted {
            // If multiple boundaries share the same time, process Ends before Starts. That's
            // because if one event ends at the same time another event starts, we do not consider
            // these events to be overlapping.
            $0.time < $1.time || $0.time == $1.time && $0.type == .End
        }

    var currentOverlap = 0
    var maxOverlap = 0

    for boundary in boundaries {
        currentOverlap += boundary.type == .Start ? 1 : -1
        maxOverlap = max(currentOverlap, maxOverlap)
    }

    return maxOverlap
}

File: EPI/14.5/typescript/14.5.ts
export interface Event {
    start: number;
    end: number;
}

interface Boundary {
    type: 'start' | 'end';
    time: number;
}

export function maxConcurrentEvents(events: Event[]): number {
    const boundaries = flatten(events.map(getBoundaries));
    boundaries.sort(compareBoundaries);

    let concurrent = 0;
    let max = 0;

    for (const b of boundaries) {
        concurrent += (b.type === 'start' ? 1 : -1);
        max = Math.max(max, concurrent);
    }

    return max;
}

const getBoundaries = (event: Event) => [
    {type: 'start', time: event.start},
    {type: 'end', time: event.end},
];

// If event A ends at the same time as event B starts, we consider these two events to not overlap.
// This means end boundaries must be placed before start boundaries with identical times.
const compareBoundaries = (a: Boundary, b: Boundary) =>
    a.time < b.time || a.time === b.time && a.type === 'end' ? -1 : 1;

const flatten = <T>(a: T[][]) => Array.prototype.concat(...a);
